dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java

```diff
@@ -47,35 +59,119 @@
     @VisibleForTesting
     BlacklistingChannel(LimitedChannel delegate, Duration duration, Ticker ticker) {
         this.delegate = delegate;
-        this.isBlacklisted = Caffeine.newBuilder()
-                .maximumSize(1)
-                .expireAfterWrite(duration)
-                .ticker(ticker)
-                .build();
+        this.duration = duration;
+        this.ticker = ticker;
+        this.perEndpointBlacklistState =
+                Caffeine.newBuilder().maximumSize(1000).ticker(ticker).build();
```

Comment:
per-endpoint state feels a bit odd here. At the channel level I think we'd want to track health of the channel as a whole. I wonder if we could invert the relationship between channel and Endpoint to allow endpoints to be wrapped similarly to conjure-undertow?

Reply:
Is this something that you want to nail down before moving forwards with this PR? (We've already got a `Cache<Endpoint, Limiter>` in our concurrency limiters...)

This BlacklistingChannel is a key ingredient for an alternative node selection strategy I've been working on, but I'd really like to land this PR as it's not super usable as is.